### PR TITLE
Fixing `AnalyzeCommandBase` and `MultithreadedAnalyzeCommandBase` artifacts generation

### PR DESCRIPTION
**Pull request by eddynaka**
# Description
The single/multi-threaded command base was generating all the artifacts even if we were not producing results.

# Test
Testing multiple files with and without the hashes flag and checking if artifacts were produced.

This PR has a dependency on this: https://github.com/microsoft/sarif-sdk/pull/2434, which will reduce the number of files to be analyzed once we merge it. That PR is just rebaselining our test files, preventing new changes from receiving all the changes we are seeing in this.

### CODE REVIEW

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 1]
```diff
@@ -307,6 +308,11 @@ private void LogCachingLogger(TContext rootContext, IAnalysisLogger logger, TCon
 
             if (results?.Count > 0)
             {
+                _run?.GetFileIndex(new ArtifactLocation { Uri = context.TargetUri },
+                                   addToFilesTableIfNotPresent: _persistArtifacts,
+                                   dataToInsert: _dataToInsert,
+                                   hashData: context.Hashes);
```
Comment (eddynaka) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:314]: we should only persist if:
1. we have results
2. if we have hashes/textfiles/binaryfiles flag enabled

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 2]
```diff
@@ -485,10 +491,6 @@ private async Task<bool> HashAsync()
                             _hashToFilesMap[hashData.Sha256] = paths;
                         }
 
-                        _run?.GetFileIndex(new ArtifactLocation { Uri = context.TargetUri },
-                                           dataToInsert: _dataToInsert,
-                                           hashData: hashData);
```
Comment (eddynaka) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:490]: this will prevent us from generating artifacts for all the files

File: src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs [hunk 3]
```diff
@@ -1165,7 +1165,7 @@ public void AnalyzeCommandBase_AutomationDetailsTests()
             }
         }
 
-        [Fact(Timeout = 5000)]
+        [Fact(Timeout = 5000, Skip = "Artifacts will be different while we don't fix SarifLogger and AnalyzeCommandBase.")]
         public void AnalyzeCommandBase_ShouldGenerateSameResultsWhenRunningSingleAndMultiThread_CoyoteTest()
```
Comment (eddynaka) [inline src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs:1169]: >AnalyzeCommandBase_ShouldGenerateSameResultsWhenRunningSingleAndMultiThread_CoyoteTest [](http://example.com/codeflow?start=20&length=86)

This test is comparing the output of a single thread vs multithreaded command base.

With this change, both will have different outputs which will be fixed in the future.

Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs:1169]: >AnalyzeCommandBase_ShouldGenerateSameResultsWhenRunningSingleAndMultiThread_CoyoteTest [](http://example.com/codeflow?start=20&length=86)

SingleAndMultithreaded

File: src/Test.UnitTests.Sarif.Multitool.Library/SuppressCommandTests.cs [hunk 4]
```diff
@@ -170,7 +170,7 @@ private static void VerifySuppressCommand(SuppressOptions options)
 
             if (options.ExpiryInDays > 0 && suppression.TryGetProperty("expiryUtc", out DateTime expiryUtc))
             {
-                expiryUtc.Should().BeCloseTo(DateTime.UtcNow.AddDays(options.ExpiryInDays));
+                expiryUtc.Should().BeCloseTo(DateTime.UtcNow.AddDays(options.ExpiryInDays), precision: 500);
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif.Multitool.Library/SuppressCommandTests.cs:173]: >500 [](http://example.com/codeflow?start=103&length=3)

used a shared constant

**Comment by michaelcfanning** [comment]
Where is my release note?

File: src/Sarif/HashUtilities.cs [hunk 5]
```diff
@@ -31,15 +31,14 @@ internal static IFileSystem FileSystem
 
         public static IDictionary<string, HashData> MultithreadedComputeTargetFileHashes(IEnumerable<string> analysisTargets, bool suppressConsoleOutput = false)
         {
-            if (analysisTargets == null) { return null; }
+            var fileToHashDataMap = new ConcurrentDictionary<string, HashData>();
+            if (analysisTargets == null) { return fileToHashDataMap; }
```
Comment (eddynaka) [inline src/Sarif/HashUtilities.cs:35]: This will prevent a NullReferenceException when we use the reerence variable from the SarifLogger

File: src/Sarif/Writers/SarifLogger.cs [hunk 6]
```diff
@@ -447,12 +447,16 @@ private void CaptureArtifact(ArtifactLocation fileLocation)
                 catch (ArgumentException) { } // Unrecognized encoding name
             }
 
+            HashData hashData = null;
+            AnalysisTargetToHashDataMap?.TryGetValue(fileLocation.Uri.OriginalString, out hashData);
```
Comment (eddynaka) [inline src/Sarif/Writers/SarifLogger.cs:451]: >AnalysisTargetToHashDataMap?.TryGetValue [](http://example.com/codeflow?start=12&length=40)

if we have the hash, let's re-use it.
this will prevent us from hashing the same file twice

File: src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs [hunk 7]
```diff
@@ -169,7 +169,7 @@ private void Analyze(TOptions options, AggregatingLogger logger)
             targets = ValidateTargetsExist(_rootContext, targets);
 
             // 5. Initialize report file, if configured.
-            InitializeOutputFile(options, _rootContext, targets);
```
Comment (eddynaka) [inline src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs:172]: >InitializeOutputFile(options, _rootContext, targets); [](http://example.com/codeflow?start=12&length=53)

we won't initialize the outputfile with targets because we don't want to emit all targets in the artifacts.

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 8]
```diff
@@ -485,12 +486,9 @@ private async Task<bool> HashAsync()
...
-
                         paths.Add(localPath);
                         context.Hashes = hashData;
+                        _pathToHashDataMap?.Add(localPath, hashData);
                     }
```
Comment (eddynaka) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:492]: Updating this will enable us to use the code from SarifLogger to create the artifacts

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/BaselineOption/ExpectedOutputs/TEST1001.ValidateWithBaseline.sarif [hunk 9]
```diff
@@ -97,15 +97,6 @@
...
       ],
-      "artifacts": [
-        {
-          "location": {
-            "uri": "FunctionalTestOutput.BaselineOption/ToBeValidated.sarif",
-            "uriBaseId": "TEST_DIR",
-            "index": 0
-          }
-        }
-      ],
```
Comment (eddynaka) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/BaselineOption/ExpectedOutputs/TEST1001.ValidateWithBaseline.sarif:108]: this will go away, since we are not using any flag to emit artifacts.

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/BaselineOption/ExpectedOutputs/TEST1001.ValidateWithBaseline.sarif [hunk 10]
```diff
@@ -59,14 +59,14 @@
                   "text": "{0}: The tool '{1}' does not provide 'informationUri'. This property helps the developer responsible for addessing a result by providing a way to learn more about the tool."
                 }
               },
-              "helpUri": "http://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html"
+              "name": "ProvideToolProperties"
             },
             {
               "id": "SARIF2012",
-              "name": "ProvideRuleProperties",
```
Comment (eddynaka) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/BaselineOption/ExpectedOutputs/TEST1001.ValidateWithBaseline.sarif:66]: >name [](http://example.com/codeflow?start=15&length=4)

the order of the properties were enforced by shaopeng in a previous PR.

File: src/Test.UnitTests.Sarif.Multitool.Library/SuppressCommandTests.cs (hunk 4 above)
Comment (eddynaka) [reply src/Test.UnitTests.Sarif.Multitool.Library/SuppressCommandTests.cs:173]: done!

File: src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs (hunk 3 above)
Comment (eddynaka) [reply src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs]: done!

Comment (eddynaka) [reply src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs]: already fixed in this pr

File: src/ReleaseHistory.md [hunk 11]
```diff
@@ -2,6 +2,7 @@
 
 ## Unreleased
 
+* BUGFIX: Fix `AnalyzeCommandBase` and `MultithreadedAnalyzeCommandBase` from outputting all artifacts to SARIF even if no results were produced when Hashes is enabled. [#2433](https://github.com/microsoft/sarif-sdk/pull/2433)
```
Comment (michaelcfanning) [inline src/ReleaseHistory.md:5]: >BUGFIX [](http://example.com/codeflow?start=2&length=6)

* BREAKING: `AnalyzeCommandBase` previously persisted all scan target artifacts to SARIF logs rather than only persisting artifacts referenced by an analysis result, when an option to persist hashes, text file or binary information was set. `MultithreadedAnalyzeCommandBase` previously persisted all scan targets artifacts to SARIF logs in cases when hash insertion was eenabled rather than only persisting artifacts referenced by an analysis result.

File: src/Sarif/HashUtilities.cs [hunk 12]
```diff
@@ -31,15 +31,14 @@ internal static IFileSystem FileSystem
 
         public static IDictionary<string, HashData> MultithreadedComputeTargetFileHashes(IEnumerable<string> analysisTargets, bool suppressConsoleOutput = false)
         {
-            if (analysisTargets == null) { return null; }
```
Comment (michaelcfanning) [inline src/Sarif/HashUtilities.cs:34]: >null [](http://example.com/codeflow?start=50&length=4)

Please revert this change. You should be careful about allocating objects to indicate no results. Why not just have the caller handle null? This code succeeded previously, so you probably have only just introduced the requirement for this to be non-null.

File: src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs (hunk 7 above)
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs:172]: >targets [](http://example.com/codeflow?start=56&length=7)

Revert this change please it's unnecessarily breaking customers.

File: src/Sarif/HashUtilities.cs (hunk 12 above)
Comment (eddynaka) [reply src/Sarif/HashUtilities.cs:34]: reverted

File: src/Sarif/OptionallyEmittedData.cs [hunk 13]
```diff
@@ -80,6 +80,9 @@ public enum OptionallyEmittedData : int
         // Enrich SARIF log with git blame information
         GitBlameInformation = 0x1000,
 
+        // Enrich with artifacts only.
+        Artifacts = 0x2000,
```
Comment (eddynaka) [inline src/Sarif/OptionallyEmittedData.cs:84]: >Artifacts [](http://example.com/codeflow?start=8&length=9)

To keep the same behavior as before, I introduced a new flag that will be able to emit the artifacts only

File: src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs (hunk 7 above)
Comment (eddynaka) [reply src/Sarif.Driver/Sdk/AnalyzeCommandBase.cs:172]: This API isn't exposed to customers. Only tools that use it will see a difference.

File: src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2012.ProvideRuleProperties_Invalid.sarif [hunk 14]
```diff
@@ -120,7 +120,7 @@
                   "index": 0
                 },
                 "region": {
-                  "startLine": 22,
```
Comment (eddynaka) [inline src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2012.ProvideRuleProperties_Invalid.sarif:123]: >22 [](http://example.com/codeflow?start=31&length=2)

we faced a similar issue in this PR:
[https://github.com/microsoft/sarif-sdk/pull/2420/files#diff-277d16bd3a7355417dd787cfc0961db73b7f32e73ad49650468d4f4bb6e5c477](https://github.com/microsoft/sarif-sdk/pull/2420/files#diff-277d16bd3a7355417dd787cfc0961db73b7f32e73ad49650468d4f4bb6e5c477)

We should investigate this.

Comment (michaelcfanning) [reply src/Test.FunctionalTests.Sarif/TestData/Multitool/ValidateCommand/ExpectedOutputs/SARIF2012.ProvideRuleProperties_Invalid.sarif:123]: Yes, let's open an issue on it.

**Review by michaelcfanning — APPROVED** [review]
:shipit:
